### PR TITLE
Add Benchmarck.Net profiling project

### PR DESCRIPTION
**Pull request by LucGenetier**
Add Benchmarck.Net profiling project to measure performance of key PowerFx APIs.
This console app needs to be compiled in release mode and will produce results like the example below.
Notice that these results will depend on the machine used.

| Method |   N |         Mean |      Error |     StdDev |          Min |           Q1 |           Q3 |          Max |
|------- |---- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------------:|
|  Parse |   1 |     5.634 us |  0.0343 us |  0.0304 us |     5.579 us |     5.625 us |     5.652 us |     5.677 us |
|  Check |   1 |    56.129 us |  1.0138 us |  0.8465 us |    55.568 us |    55.673 us |    56.153 us |    58.825 us |
|   Eval |   1 |    57.456 us |  0.3150 us |  0.2947 us |    56.932 us |    57.324 us |    57.697 us |    57.764 us |
|  Parse |   2 |    12.029 us |  0.0651 us |  0.0577 us |    11.950 us |    11.996 us |    12.065 us |    12.159 us |
|  Check |   2 |   114.498 us |  1.0578 us |  0.9377 us |   112.682 us |   113.873 us |   115.152 us |   116.451 us |
|   Eval |   2 |   112.719 us |  0.9399 us |  0.8792 us |   111.661 us |   111.926 us |   113.149 us |   114.603 us |
|  Parse |   5 |    30.931 us |  0.5996 us |  0.8208 us |    29.922 us |    30.253 us |    31.317 us |    32.828 us |
|  Check |   5 |   283.300 us |  2.9594 us |  2.7682 us |   279.459 us |   281.388 us |   284.453 us |   288.856 us |
|   Eval |   5 |   270.327 us |  2.9469 us |  2.6123 us |   265.346 us |   268.750 us |   271.924 us |   275.122 us |
|  Parse |  10 |    59.969 us |  0.2200 us |  0.1951 us |    59.685 us |    59.838 us |    60.144 us |    60.262 us |
|  Check |  10 |   553.311 us |  3.4061 us |  3.0194 us |   546.914 us |   551.101 us |   554.823 us |   558.592 us |
|   Eval |  10 |   527.913 us |  2.8641 us |  2.5390 us |   523.217 us |   526.081 us |   529.657 us |   532.359 us |
|  Parse |  20 |   126.709 us |  2.5166 us |  5.5239 us |   119.027 us |   120.950 us |   131.278 us |   136.713 us |
|  Check |  20 | 1,182.082 us | 23.3672 us | 37.0628 us | 1,121.656 us | 1,151.371 us | 1,213.736 us | 1,245.586 us |
|   Eval |  20 | 1,023.810 us |  7.3008 us |  6.8292 us | 1,016.347 us | 1,017.929 us | 1,028.714 us | 1,037.119 us |
|  Parse |  50 |   304.723 us |  2.3317 us |  2.0670 us |   301.125 us |   303.259 us |   306.117 us |   307.844 us |
|  Check |  50 | 2,776.818 us |  6.5057 us |  6.0854 us | 2,767.250 us | 2,772.776 us | 2,781.323 us | 2,786.738 us |
|   Eval |  50 | 2,688.011 us | 52.6876 us | 86.5673 us | 2,578.290 us | 2,609.024 us | 2,761.661 us | 2,836.030 us |
|  Parse | 100 |   628.569 us |  3.7187 us |  3.2965 us |   623.051 us |   626.312 us |   630.026 us |   635.731 us |
|  Check | 100 | 5,537.963 us | 19.3509 us | 17.1540 us | 5,514.151 us | 5,525.402 us | 5,549.316 us | 5,577.848 us |
|   Eval | 100 | 5,135.896 us | 60.8737 us | 56.9413 us | 5,056.144 us | 5,097.930 us | 5,167.779 us | 5,263.991 us |

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            engine = new Engine(powerFxConfig);
+            parserOptions = new ParserOptions() { AllowsSideEffects = true, Culture = new CultureInfo("en-US") };
+            recalcEngine = new RecalcEngine(powerFxConfig);
+        }
+
+        [Params(1, 2, 5, 10, 20, 50, 100)]
+        public int N { get; set; }
+
+        [Benchmark]
+        public ParseResult Parse()
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs:37]: >Parse [](http://example.com/codeflow?start=27&length=5)

Could we add one for just Lex?

**Comment by MikeStall** [comment]
>     public enum Features : int

This makes me wonder ... what happens when we run out of flags? int only has 32 bits....  to be safe, we shouldn't recycle bits. 

The alternative is a class with member flags.

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:9 in 490a9df. [](commit_id = 490a9df16d1c6ec6731d95546156c73b635c4f2a, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    public class PerformanceTests
+    {
+        [Fact]
+        public void PerformanceTest()
+        {
+            var pt = new PerformanceTest1();
+            pt.GlobalSetup();
+            pt.N = 20;
+
+            for (var i = 0; i < 10000; i++)
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs:17]: >for (var i = 0; i < 10000; i++) [](http://example.com/codeflow?start=12&length=31)

why bother looping here?

Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs:17]: to run VS profiler

**Comment by LucGenetier** [comment]
>     public enum Features : int

ulong, then BigInteger or we'll create a List<string>

---
In reply to: [1341118988](https://github.com/microsoft/Power-Fx/pull/887#issuecomment-1341118988) [](http://example.com/codeflow?ancestors=1341118988,1341118988)

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:9 in 490a9df. [](commit_id = 490a9df16d1c6ec6731d95546156c73b635c4f2a, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Performance.Tests/Program.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+namespace Microsoft.PowerFx.Performance.Tests
+{
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            /*
+             * 
+             *       --> This program needs to be run ELEVATED
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Performance.Tests/Program.cs:15]: >--> [](http://example.com/codeflow?start=21&length=3)

does this also need to be run in Release?

**Review by Shpakh-MSFT — DISMISSED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs (hunk 2 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs:17]: Sometimes convenient to run VS profiles, hence this test

File: src/tests/Microsoft.PowerFx.Performance.Tests/Program.cs (hunk 3 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Performance.Tests/Program.cs:15]: Yes, this project isn't compiled in Debug

**Review by Shpakh-MSFT — DISMISSED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs (hunk 2 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Performance.Tests/UnitTests/PerformanceTests.cs:17]: Just get this from env variable. In CI, defaults to 1.  If we want to profile, can set to 10k.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Pull request merged**